syntax = "proto3";

message CreateVectorRequest {
    repeated float vector = 1;
}

message SearchVectorRequest {
    repeated float query = 1;
    int32 top_k = 2;
}

message CreateVectorResponse {
    string id = 1;
}

message SearchVectorResponse {
    repeated SearchVectorResult results = 1;
}

message SearchVectorResult {
    string id = 1;
    float similarity = 2;
}

message GetVectorByIdRequest {
    string id = 1;
}

message GetVectorByIdResponse {
    repeated float query = 1;
}

service ViperDB {
    rpc CreateVector(CreateVectorRequest) returns (CreateVectorResponse);
    rpc SearchVector(SearchVectorRequest) returns (SearchVectorResponse);
    rpc GetVectorById(GetVectorByIdRequest) returns (GetVectorByIdResponse);
}